cmd/deleteaccount.go

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright 2019 The NATS Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	cli "github.com/nats-io/cliprompts/v2"
+	"github.com/nats-io/jwt"
+	"github.com/nats-io/nkeys"
+	"github.com/nats-io/nsc/cmd/store"
+	"github.com/spf13/cobra"
+)
+
+func createDeleteAccountCmd() *cobra.Command {
+	var params DeleteAccountParams
+	cmd := &cobra.Command{
+		Use:   "account",
+		Short: "Delete an account and associated users",
+		Args:  cobra.MaximumNArgs(1),
+		Example: `nsc delete account -n name
+nsc delete account -i
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return RunAction(cmd, args, &params)
+		},
+	}
+
+	cmd.Flags().StringVarP(&params.AccountContextParams.Name, "name", "n", "", "name of account to delete")
+	cmd.Flags().BoolVarP(&params.revoke, "revoke", "R", true, "revoke users before deleting")
+	cmd.Flags().BoolVarP(&params.rmNkeys, "rm-nkey", "D", false, "delete user keys")
+	cmd.Flags().BoolVarP(&params.rmCreds, "rm-creds", "C", false, "delete users creds")
+	cmd.Flags().BoolVarP(&params.force, "force", "F", false, "managed accounts must supply --force")
+
+	return cmd
+}
+
+func init() {
+	deleteCmd.AddCommand(createDeleteAccountCmd())
+}
+
+type DeleteAccountParams struct {
+	AccountContextParams
+	SignerParams
+	ac      *jwt.AccountClaims
+	force   bool
+	revoke  bool
+	rmCreds bool
+	rmNkeys bool
+	users   []string
+}
+
+func (p *DeleteAccountParams) SetDefaults(ctx ActionCtx) error {
+	p.AccountContextParams.Name = NameFlagOrArgument(p.AccountContextParams.Name, ctx)
+	if err := p.AccountContextParams.SetDefaults(ctx); err != nil {
+		return err
+	}
+	p.SignerParams.SetDefaults(nkeys.PrefixByteOperator, true, ctx)
+
+	return nil
+}
+
+func (p *DeleteAccountParams) PreInteractive(ctx ActionCtx) error {
+	var err error
+	if err = p.AccountContextParams.Edit(ctx); err != nil {
+		return err
+	}
+	p.revoke, err = cli.Confirm("revoke all account users before deleting", true)
```

Comment:
When we revoke users we formally need the account to still be active to hold the revocations. Should this be named slightly different maybe?

Reply:
The delete account process does the following:
- Revokes all users
- Removes all account signing keys
- Sets the expiration of the account JWT to 1m into the future
- Pushes the account JWT to the account server (if any is specified in by the operator)
- Deletes the user JWTs
- Deletes the local account JWT
- Deletes user creds if specified
- Deletes the user, account signing keys, and account nkeys if specified

The remote system will have a record, but any user signed account will not be valid. If the nkeys to the account are preserved, it may be possible to re-issue (even sync the remote one), and start over.

Reply:
> "Deleting" an account is not meant to be compatible with our contemplated bearer authorization, right?

Deleting an account means invalidating an entity that issues users. If you have a bearer token (an user) that was issued by an account and that account is not valid in the system, the user is also not valid.

Reply:
Agree, should we have the operator revoke the account?

Reply:
I see... this essentially revokes the key which would be referenced by `kid` in the bearer token header.

The ability to revoke a key obviously makes sense, but under the bearer model it seems odd that a user's JWTs exist on the server and can be "deleted."

Reply:
Server never holds onto a user jwt, so no worries there.